clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/KafkaStoreIngestionService.java

```diff
@@ -1054,6 +1073,76 @@ public AdminResponse getConsumptionSnapshots(String topicName, ComplementSet<Int
     return response;
   }
 
+  /**
+   * Return the metadata information for the given store. The data is retrieved from its respective repositories which
+   * originate from the VeniceServer.
+   * @param storeName
+   * @return {@link MetadataResponse} object that holds all the information required for answering a server metadata
+   * fetch request.
+   */
+  @Override
+  public MetadataResponse getMetadata(String storeName) {
+    MetadataResponse response = new MetadataResponse();
+    try {
+      Store store = metadataRepo.getStoreOrThrow(storeName);
+
+      List<Integer> versions = new ArrayList<>();
+      for (Version version: store.getVersions()) {
+        version.getPartitionerConfig();
+        versions.add(version.getNumber());
+      }
+      Map<CharSequence, CharSequence> partitionerParams =
+          new HashMap<>(store.getPartitionerConfig().getPartitionerParams());
+      VersionProperties versionProperties = new VersionProperties(
+          store.getCurrentVersion(),
+          versions,
+          store.getCompressionStrategy().getValue(),
+          store.getPartitionCount(),
+          store.getPartitionerConfig().getPartitionerClass(),
+          partitionerParams,
+          store.getPartitionerConfig().getAmplificationFactor());
+
+      Map<CharSequence, CharSequence> keySchema = Collections.singletonMap(
+          String.valueOf(schemaRepo.getKeySchema(storeName).getId()),
+          schemaRepo.getKeySchema(storeName).getSchema().toString());
+      Map<CharSequence, CharSequence> valueSchemas = new HashMap<>();
+      for (SchemaEntry schemaEntry: schemaRepo.getValueSchemas(storeName)) {
+        String valueSchemaStr = schemaEntry.getSchema().toString();
+        int valueSchemaId = schemaRepo.getValueSchemaId(storeName, valueSchemaStr);
+        valueSchemas.put(String.valueOf(valueSchemaId), valueSchemaStr);
+      }
+
+      Map<CharSequence, List<CharSequence>> routingInfo = new HashMap<>();
+      for (String resource: routingRepository.getResourceAssignment().getAssignedResources()) {
+        if (resource.endsWith("v" + store.getCurrentVersion())) {
+          for (Partition partition: routingRepository.getPartitionAssignments(resource).getAllPartitions()) {
+            List<CharSequence> instances = new ArrayList<>();
+            for (Instance instance: partition.getInstanceToStateMap().keySet()) {
+              instances.add(instance.toString());
```

Comment:
Should we only add instances that are actually ready to serve? i.e. CV state is COMPLETED. I don't see any additional info is passed to the client. If we add all instances regardless of their state then we might be routing request to a replica that we already know at some point that it's not ready.

Reply:
Totally agree, will make the change to only add instances in COMPLETED state